src/ray/common/scheduling/scheduling_alg.h

```diff
@@ -0,0 +1,147 @@
+#ifndef RAY_COMMON_SCHEDULING_SCHEDULING_H
+#define RAY_COMMON_SCHEDULING_SCHEDULING_H
+
+#include <iostream>
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+
+/// Maximum number of predefined resources.
+#define NUM_PREDIFINED_RESOURCES 4
+/// List of predefined resources.
+/// The number of elements in enum should be equal to NUM_PREDIFINED_RESOURCES.
+enum PredefinedResources {CPU = 0, MEM = 1, GPU = 2, TPU = 3};
+
+/// Total and available capacity of a resource instance.
+struct ResourceCapacity {
+  int64_t total;
+  int64_t available;
+};
+
+/// Resource request.
+struct ResourceReq {
+  /// Amount of resource being requested.
+  int64_t demand;
+  /// Specify whether the request is soft or hard.
+  /// If hard, the entire request is denied if the demand exceeds the resource
+  /// availability. Otherwise, the request can be still be granted.
+  bool soft;
+};
+
+/// Resource request, including resource ID. This is used for custom resources.
+struct ResourceReqWithId {
+  /// Resource ID.
+  int64_t id;
+  /// Resource request.
+  ResourceReq req;
+};
+
+
+/// All resources associated with a node.
+struct NodeResources {
+  /// Available and total capacities for predefined resources.
+  std::vector<ResourceCapacity> capacities;
+  /// Map containing custom resources. The key of each entry represents the
+  /// custom resource ID.
+  std::unordered_map<int64_t, ResourceCapacity> custom_resources;
+};
+
+/// Task request.
+struct TaskReq {
+  /// List of predefined resources.
+  std::vector<ResourceReq> predefined_resources;
+  /// List of custom resources.
+  std::vector<ResourceReqWithId> custom_resources;
+  /// List of placement hints. A placement hint is a node on which
+  /// we desire to run this task. This is a soft constraint in that
+  /// the task will run on a different node in the cluster, if none of the
+  /// nodes in this list can schedule this task.
+  std::unordered_set<int64_t> placement_hints;
+};
+
+/// Class encapsulating the cluster resources and the logic to assign
+/// tasks to nodes based on the task's constraints and the available
+/// resources at those nodes.
+class ClusterResources {
```

Comment:
Seems like `ClusterResourceScheduler` would make more sense, it doesn't just hold the resources

Reply:
also please rename `scheduling_alg.{cc,h}` to whatever you land on here, it doesn't really fit what's in the file at the moment